src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -5493,14 +5493,35 @@ class StubGenerator: public StubCodeGenerator {
   }
 
   // r0 = input (float16)
-  // v0 = result (float)
+  // v0 = result (float or double)
   // v1 = temporary float register
-  address generate_float16ToFloat() {
+  address generate_float16ToFP(BasicType bt) {
     __ align(CodeEntryAlignment);
-    StubCodeMark mark(this, "StubRoutines", "float16ToFloat");
+    StubCodeMark mark(this, "StubRoutines", "float16 to float/double");
     address entry = __ pc();
     BLOCK_COMMENT("Entry:");
-    __ flt16_to_flt(v0, r0, v1);
+    switch (bt) {
+      case T_FLOAT:  __ flt16_to_flt(v0, r0, v1, T_FLOAT);  break;
+      case T_DOUBLE: __ flt16_to_flt(v0, r0, v1, T_DOUBLE); break;
+      default: ShouldNotReachHere();
+    }
```

Comment:
Ok, I re-visited this, so conversion stubs for constant folding call direct FP16 to INT/LONG/DOUBLE instructions. 

This looks reasonable.  Though constant folding is something that happens at compile time so it may not result in any runtime penalty even if we remove the stubs and directly cast to target type after hf2f stub.